File: c++/geometry.cpp
#define _USE_MATH_DEFINES
#include "geometry.h"
#include <iostream>
#include <vector>
#include <tuple>
#include <math.h>
using namespace std;

//Shape
const vector<tuple<double, double>>& Shape::getVertices() {
    return vertices;
}

tuple<double, double> Shape::getCenter() {
    return *centerPtr;
}

void Shape::print() {
    for (auto vertex : vertices) {
        printf("X: %g | Y: %g\n", get<0>(vertex), get<1>(vertex));
    }
}

Shape::~Shape() {
    vertices.clear();
    vertices.shrink_to_fit();

    delete centerPtr;
    centerPtr = nullptr;
}

//Circle
Circle::Circle(tuple<double, double> pos, double radius, int resolution) {
    centerPtr = &pos;
    this->radius = radius;

    for (int i = 0; i < resolution; i++) {
        double angle = (double)i / resolution * 2 * M_PI;

        tuple<double, double> vertex(
            get<0>(pos) + radius * cos(angle),
            get<1>(pos) + radius * sin(angle));

        vertices.push_back(vertex);
    }
}

double Circle::getLength() {
    return 2 * M_PI * radius;
}

double Circle::getArea() {
    return M_PI * pow(radius,2);
}

//Square
Square::Square(tuple<double, double> in_vertices[4]) : Rectangle(in_vertices) {}

double Square::getLength() {
    return sqrt(pow(get<0>(vertices[0]),2) + pow(get<1>(vertices[0]),2)) * 4;
}

//Rectangle
Rectangle::Rectangle(tuple<double, double> in_vertices[4]) {
    double x = 0, y = 0;

    for (int i = 0; i < 4; i++) {
        vertices.push_back(in_vertices[i]);

        x += get<0>(in_vertices[i]);
        y += get<1>(in_vertices[i]);
    }

    centerPtr = new tuple<double, double> (x/4, y/4);
}

double Rectangle::getLength() {
    double length = 0;

    for (int i = 0; i < 2; i++) {
        double x = 0, y = 0;

        x = get<0>(vertices[i + 1]) - get<0>(vertices[i]);
        y = get<1>(vertices[i + 1]) - get<1>(vertices[i]);

        length += sqrt(pow(x,2) + pow(y,2));
    }

    return length * 2;
}

double Rectangle::getArea() {
    double x = get<0>(vertices[1]) - get<0>(vertices[0]);
    double y = get<1>(vertices[1]) - get<1>(vertices[0]);

    double area = sqrt(pow(x,2) + pow(y,2));

    x = get<0>(vertices[2]) - get<0>(vertices[1]);
    y = get<1>(vertices[2]) - get<1>(vertices[1]);

    return area * sqrt(pow(x,2) + pow(y,2));
}

File: python/geometry.py
from math import pi, cos, sin, sqrt

class Shape:
    def __init__(self):
        self.__vertices = []
        self.__center = (0,0)
    def __str__(self):
        res = ""
        for vertex in self.__vertices:
            x, y = vertex
            res += f"X: {x} | Y: {y}\n"
        return res
    def getVertices(self): return self.__vertices
    def getCenter(self): return self.__center
    def __setVertices(self, newValue): self.__vertices = newValue
    def __setCenter(self, newValue): self.__center = newValue
    Vertices = property(lambda x : x.__vertices, __setVertices)
    Center = property(lambda x : x.__center, __setCenter)

class Circle (Shape):
    def __init__(self, pos, radius, resolution):
        Shape.__init__(self)
        self.Center = pos
        self.__radius = radius
        vertices = []
        for i in range(0, resolution):
            angle = i/resolution * 2 * pi

            vertices.append((pos[0] + radius * cos(angle),
                                    pos[1] + radius * sin(angle)))
        self.Vertices = vertices
    def getRadius(self): return self.__radius
    def getArea(self): return pi * self.__radius**2
    def getLength(self): return 2 * pi * self.__radius

class Rectangle (Shape):
    def __init__(self, vertices):
        Shape.__init__(self)
        self.Vertices = vertices
        x, y = 0, 0
        for vertex in vertices:
            x += vertex[0]
            y += vertex[1]
        self.Center = (x,y)
    def getArea(self):
        x = self.Vertices[1][0] - self.Vertices[0][0]
        y = self.Vertices[1][1] - self.Vertices[0][1]

        area = sqrt(x**2 + y**2)

        x = self.Vertices[2][0] - self.Vertices[1][0]
        y = self.Vertices[2][1] - self.Vertices[1][1]

        return area * sqrt(x**2 + y**2)
    def getLength(self):
        length = 0

        for i in range(0,2):
            x = self.Vertices[i + 1][0] - self.Vertices[i][0]
            y = self.Vertices[i + 1][1] - self.Vertices[i][1]

            length += sqrt(x**2 + y**2)
        return length * 2
    
class Square (Rectangle):
    def __init__(self, vertices):
        Rectangle.__init__(self, vertices)
    def getLength(self):
        return sqrt(self.Vertices[0][0]**2 + self.Vertices[0][1]**2) * 4